common/flatpak-installation.c

```diff
@@ -2962,67 +3097,46 @@ flatpak_installation_list_unused_refs (FlatpakInstallation *self,
+      g_autoptr(GFile) user_base_dir = flatpak_get_user_base_dir_location ();
 
-      if (arch != NULL && strcmp (parts[2], arch) != 0)
-        continue;
+      if (g_file_query_exists (user_base_dir, cancellable))
+        {
+          g_autoptr(FlatpakDir) user_dir = flatpak_dir_get_user ();
+          g_auto(GStrv) user_app_refs = NULL;
+          g_auto(GStrv) user_runtime_refs = NULL;
+          g_autoptr(GHashTable) user_runtimes = g_hash_table_new_full (g_str_hash, g_str_equal, NULL, NULL);
 
-      deploy = flatpak_dir_load_deployed (dir, ref, NULL, NULL, NULL);
-      if (deploy == NULL)
-        continue;
+          if (!flatpak_dir_list_refs (user_dir, "app", &user_app_refs, cancellable, error))
+            return NULL;
 
-      origin = flatpak_dir_get_origin (dir, ref, NULL, NULL);
-      if (origin == NULL)
-        continue;
+          find_used_refs_for_apps (user_dir, dir, user_app_refs, arch, used_runtimes, used_refs);
 
-      find_used_refs (dir, used_refs, ref, origin);
+          if (!flatpak_dir_list_refs (user_dir, "runtime", &user_runtime_refs, cancellable, error))
+            return NULL;
 
-      metakey = flatpak_deploy_get_metadata (deploy);
-      runtime = g_key_file_get_string (metakey, "Application", "runtime", NULL);
-      if (runtime)
-        g_hash_table_add (used_runtimes, g_steal_pointer (&runtime));
+          for (i = 0; user_runtime_refs[i] != NULL; i++)
+            {
+              const char *ref = user_runtime_refs[i];
+              g_assert (g_str_has_prefix (ref, "runtime/"));
+              g_hash_table_add (user_runtimes, (char *)ref + strlen ("runtime/"));
+            }
 
-      sdk = g_key_file_get_string (metakey, "Application", "sdk", NULL);
-      if (sdk)
-        g_hash_table_add (used_runtimes, g_steal_pointer (&sdk));
+          find_used_refs_for_runtimes (user_dir, dir, user_runtimes, used_refs);
+        }
     }
 
-  GLNX_HASH_TABLE_FOREACH (used_runtimes, const char *, runtime)
-  {
-    g_autofree char *runtime_ref = g_strconcat ("runtime/", runtime, NULL);
-    g_autoptr(FlatpakDeploy) deploy = NULL;
-    g_autofree char *origin = NULL;
-    g_autofree char *sdk = NULL;
-    g_autoptr(GKeyFile) metakey = NULL;
-
-    deploy = flatpak_dir_load_deployed (dir, runtime_ref, NULL, NULL, NULL);
-    if (deploy == NULL)
-      continue;
-
-    origin = flatpak_dir_get_origin (dir, runtime_ref, NULL, NULL);
-    if (origin == NULL)
-      continue;
-
-    find_used_refs (dir, used_refs, runtime_ref, origin);
-
-    metakey = flatpak_deploy_get_metadata (deploy);
-    sdk = g_key_file_get_string (metakey, "Runtime", "sdk", NULL);
-    if (sdk)
-      {
-        g_autofree char *sdk_ref = g_strconcat ("runtime/", sdk, NULL);
-        g_autofree char *sdk_origin = flatpak_dir_get_origin (dir, sdk_ref, NULL, NULL);
-        if (sdk_origin)
-          find_used_refs (dir, used_refs, sdk_ref, sdk_origin);
-      }
-  }
+  find_used_refs_for_runtimes (dir, NULL, used_runtimes, used_refs);
```

Comment:
Maybe we can move this call before the if (!is_user) branch above which does the user + system dir version of this call. Its not super clear that we do both when the order is mixed up like this.

Reply:
I guess we need to scan the apps first, but maybe we can move the second _for_runtimes call after this.